sympy/sets/tests/test_sets.py

```diff
@@ -1597,6 +1600,50 @@ def test_DisjointUnion_len():
     assert len(DisjointUnion(S.EmptySet, S.EmptySet, FiniteSet(x, y, z), S.EmptySet)) == 3
     raises(ValueError, lambda: len(DisjointUnion(Interval(0, 1), S.EmptySet)))
 
+def test_SetKind_ProductSet():
+    p = ProductSet(FiniteSet(Matrix([1, 2])), FiniteSet(Matrix([1, 2])))
+    mk = MatrixKind(NumberKind)
+    k = SetKind(TupleKind(mk, mk))
+    n =  ProductSet(Interval(1, 2), FiniteSet(Matrix([1, 2]))).kind
+    q = SetKind(TupleKind(NumberKind, mk))
+    assert p.kind is k
+    assert n is q
+
+def test_SetKind_Interval():
+    assert Interval(1, 2).kind is SetKind(NumberKind)
+
+def test_SetKind_EmptySet_UniversalSet():
+    assert S.UniversalSet.kind is SetKind(UndefinedKind)
+    assert EmptySet.kind is SetKind()
+
+def test_SetKind_FiniteSet():
+    assert FiniteSet(1, Matrix([1, 2])).kind is SetKind(UndefinedKind)
+    assert FiniteSet(1, 2).kind is SetKind(NumberKind)
+
+def test_SetKind_Unions():
+    assert Union(FiniteSet(Matrix([1, 2])), Interval(1, 2)).kind is SetKind(UndefinedKind)
+    assert Union(Interval(1, 2), Interval(1, 7)).kind is SetKind(NumberKind)
+
+def test_SetKind_evaluate_False():
+    one = Union({1}, EmptySet, evaluate=False).kind
+    one_ = Union({1}, EmptySet).kind
+    two = Union(Interval(1, 2), EmptySet, evaluate=False).kind
+    two_ = Union(Interval(1, 2), EmptySet).kind
+    three = Intersection({1}, S.UniversalSet,evaluate=False).kind
+    three_ = Intersection({1}, S.UniversalSet).kind
+    four = Union({1}, S.UniversalSet, evaluate=False).kind
+    four_ =Union({1}, S.UniversalSet).kind
+    five = Union(Interval(1, 2), Interval(4, 5), FiniteSet(1), evaluate=False).kind
+    five_ = Union(Interval(1, 2), Interval(4, 5), FiniteSet(1)).kind
+    six = Intersection({1}, EmptySet, evaluate=False).kind
+    six_ = Intersection({1}, EmptySet).kind
+    assert one is one_
```

Comment:
It would be better to explicitly assert what the kinds are rather than just that the evaluate=False/True results are equivalent.